test/performance/odp_mem_perf.c

```diff
@@ -0,0 +1,435 @@
+	if (num_cpu == 0) {
+		num_cpu = ret;
+		test_options->num_cpu = num_cpu;
+	}
+
+	odp_barrier_init(&global->barrier, num_cpu);
+
+	return 0;
+}
+
+static int create_shm(test_global_t *global)
+{
+	odp_shm_capability_t shm_capa;
+	odp_shm_t shm;
+	void *addr;
+	uint32_t i, num_shm;
+	test_options_t *test_options = &global->test_options;
+	uint32_t num_round = test_options->num_round;
+	uint32_t num_cpu   = test_options->num_cpu;
+	uint64_t data_len  = test_options->data_len;
+	uint32_t shm_flags = test_options->shm_flags;
+	int private        = test_options->private;
+	char name[] = "mem_perf_00";
+
+	num_shm = 1;
+	if (private)
+		num_shm = num_cpu;
+
+	printf("\nMemory performance test\n");
+	printf("  num cpu          %u\n", num_cpu);
+	printf("  num rounds       %u\n", num_round);
+	printf("  data len         %" PRIu64 "\n", data_len);
+	printf("  memory footprint %" PRIu64 "\n", num_shm * data_len);
+	printf("  shm flags        0x%x\n", shm_flags);
+	printf("  num shm          %u\n", num_shm);
+	printf("  private          %i\n", private);
+
+	if (odp_shm_capability(&shm_capa)) {
+		ODPH_ERR("SHM capa failed.\n");
+		return -1;
+	}
+
+	if (shm_capa.max_size && data_len > shm_capa.max_size) {
+		ODPH_ERR("Data len too large. Maximum len is %" PRIu64 "\n", shm_capa.max_size);
+		return -1;
+	}
+
+	if (num_shm > shm_capa.max_blocks) {
+		ODPH_ERR("Too many SHM blocks. Maximum is %u\n", shm_capa.max_blocks);
+		return -1;
+	}
+
+	for (i = 0; i < num_shm; i++) {
+		name[9]  += i / 10;
+		name[10] += i % 10;
+
+		shm = odp_shm_reserve("mem_perf", data_len, ODP_CACHE_LINE_SIZE, shm_flags);
+
+		if (shm == ODP_SHM_INVALID) {
+			ODPH_ERR("SHM[%u] reserve failed.\n", i);
+			return -1;
+		}
+
+		global->shm[i] = shm;
+
+		addr = odp_shm_addr(shm);
+		if (addr == NULL) {
+			ODPH_ERR("SHM[%u] addr failed.\n", i);
+			return -1;
+		}
+
+		global->shm_addr[i] = addr;
+
+		printf("  shm addr[%u]      %p\n", i, addr);
+	}
+
+	printf("\n");
+	global->num_shm = num_shm;
+
+	return 0;
+}
+
+static int free_shm(test_global_t *global)
+{
+	uint32_t i;
+
+	for (i = 0; i < global->num_shm; i++) {
+		if (odp_shm_free(global->shm[i])) {
+			ODPH_ERR("SHM[%u] free failed.\n", i);
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+static int run_test(void *arg)
+{
+	int thr;
+	uint32_t i;
+	uint64_t nsec;
+	odp_time_t t1, t2;
+	test_thread_ctx_t *thread_ctx = arg;
+	test_global_t *global = thread_ctx->global;
+	test_options_t *test_options = &global->test_options;
+	uint32_t num_round = test_options->num_round;
+	uint64_t data_len = test_options->data_len;
+	uint64_t half_len = data_len / 2;
+	int mode = test_options->mode;
+	uint8_t *addr = thread_ctx->shm_addr;
+
+	thr = odp_thread_id();
+
+	/* Start all workers at the same time */
+	odp_barrier_wait(&global->barrier);
+
+	t1 = odp_time_local();
+
+	if (mode == 0) {
+		for (i = 0; i < num_round; i++)
+			memset(addr, thr + i, data_len);
+	} else {
+		for (i = 0; i < num_round; i++) {
+			if (i % 2 == 0)
```

Comment:
Better to change to AND opertaion in fastpath instead of %. i.e i & 1

Reply:
True. Fixed in v2.